src/main/kotlin/com/exactpro/th2/dataprocessor/zephyr/impl/cache/LRUCache.kt

```diff
@@ -16,27 +16,70 @@
 
 package com.exactpro.th2.dataprocessor.zephyr.impl.cache
 
+import mu.KotlinLogging
 import org.apache.commons.collections4.map.LRUMap
-import java.util.function.LongSupplier
+import java.time.Clock
+import java.time.Duration
+import java.time.Instant
+import java.time.LocalDate
+import java.time.LocalTime
+import java.time.ZoneOffset
+import java.time.temporal.ChronoUnit
 
 internal class LRUCache<K, V>(
     size: Int,
     private val expireAfterMillis: Long,
-    private val timeSource: LongSupplier = LongSupplier(System::currentTimeMillis),
+    private val timeSource: Clock = Clock.systemUTC(),
+    invalidateAt: LocalTime? = null,
 ) {
+    private var nextInvalidateAt: Instant
     init {
         require(size > 0) { "size must be positive but was $size" }
         require(expireAfterMillis > 0) { "expireAfterMillis must be positive but was $expireAfterMillis" }
+        nextInvalidateAt = if (invalidateAt == null) {
+            Instant.MAX
+        } else {
+            val now = timeSource.instant()
+            val next = invalidateAt.atDate(LocalDate.ofInstant(now, ZoneOffset.UTC))
+                .toInstant(ZoneOffset.UTC)
+            if (next <= now) {
+                next.plus(1, ChronoUnit.DAYS)
+            } else {
+                next
+            }
+        }
     }
-    private class TimestampedValue<T>(val value: T, val createdAt: Long)
+    private class TimestampedValue<T>(val value: T, val createdAt: Instant)
 
     private val cache = LRUMap<K, TimestampedValue<V>>(size)
 
-    operator fun get(key: K): V? = cache[key]?.takeUnless {
-        timeSource.asLong - it.createdAt > expireAfterMillis
-    }?.value
+    operator fun get(key: K): V? {
+        val now = timeSource.instant()
+        if (now >= nextInvalidateAt) {
+            invalidateAll()
+        }
+        return cache[key]?.takeUnless {
+            Duration.between(it.createdAt, now).toMillis() > expireAfterMillis
+        }?.value
+    }
 
     operator fun set(key: K, value: V) {
-        cache[key] = TimestampedValue(value, timeSource.asLong)
+        cache[key] = TimestampedValue(value, timeSource.instant())
+    }
+
+    private fun invalidateAll() {
+        cache.clear()
+        val currentInvalidateAt = nextInvalidateAt
+        var next: Instant = nextInvalidateAt
+        val now = timeSource.instant()
+        do {
+            next = next.plus(1, ChronoUnit.DAYS)
+        } while (next < now)
```

Comment:
Maybe use while ... do instead because I think we should grantee invalidation at configured time

Reply:
Could you advise what do you mean by that? This loop computes the next invalidation time based on the current time and the last invalidation time. There is no inner timer in the cache so the invalidation might not be called at the exact time. However, we just need to guarantee that if invalidation time has passed data in cache will be invalidated. And compute the next invalidation time. Also, if you look at the code you will see that this is the private method and it is invoked only if it is time to invalidate the cache. Because of that, using do-while here works fine, in my opinion.

Reply:
```
Also, if you look at the code you will see that this is the private method and it is invoked only if it is time to invalidate the cache. Because of that, using do-while here works fine, in my opinion.
```
Thank you for this point. I agree with you.
I look at this method as a separate and we have invalidateAll name and calculate new value nextInvalidateAt. If I reuse this method in another place of this class, I will not look into it and can forget that the method adds 1 day in any cases.